### PR TITLE
feat: update ICA to v3 

### PR DESCRIPTION
**Pull request by aroralanuk**
### Description

- Update ICA to accept _hookMetadata per call and per batch of calls.
- Allow ICA to accept msg.value

### Drive-by changes

- Moved IGP from MockMailbox to the InterchainAccountRouterTest contract as it's not needed in the former and doesn't break any tests

### Related issues

- Fixes https://github.com/hyperlane-xyz/hyperlane-monorepo/issues/3197

### Backward compatibility

Yes with V3

### Testing

Unit tests with focus on igp as the hook as this will be the one used in production

### CODE REVIEW

File: solidity/contracts/middleware/InterchainAccountRouter.sol [hunk 1]
```diff
@@ -1,12 +1,25 @@
 // SPDX-License-Identifier: Apache-2.0
 pragma solidity ^0.8.13;
 
+/*@@@@@@@       @@@@@@@@@
```
Comment (yorhodes) [inline solidity/contracts/middleware/InterchainAccountRouter.sol:4]: boosting those contribution numbers lol

File: solidity/contracts/middleware/InterchainAccountRouter.sol [hunk 2]
```diff
@@ -150,14 +163,16 @@ contract InterchainAccountRouter is Router {
...
      */
     function callRemote(
         uint32 _destination,
         address _to,
         uint256 _value,
-        bytes memory _data
-    ) external returns (bytes32) {
+        bytes memory _data,
+        bytes memory _hookMetadata
+    ) external payable returns (bytes32) {
```
Comment (yorhodes) [inline solidity/contracts/middleware/InterchainAccountRouter.sol:175]: can we keep these functions for backwards compatibility and just add new ones with hook metadata?

File: solidity/contracts/middleware/InterchainAccountRouter.sol [hunk 3]
```diff
@@ -388,20 +444,29 @@ contract InterchainAccountRouter is Router {
...
      */
     function callRemoteWithOverrides(
         uint32 _destination,
         bytes32 _router,
         bytes32 _ism,
-        CallLib.Call[] calldata _calls
-    ) public returns (bytes32) {
+        CallLib.Call[] calldata _calls,
+        bytes memory _hookMetadata
+    ) public payable returns (bytes32) {
```
Comment (yorhodes) [inline solidity/contracts/middleware/InterchainAccountRouter.sol:456]: ditto here?

File: solidity/contracts/middleware/InterchainAccountRouter.sol [hunk 4]
```diff
@@ -474,7 +539,33 @@ contract InterchainAccountRouter is Router {
...
+     */
+    function _dispatchMessageWithMetadata(
+        uint32 _destination,
+        bytes32 _router,
+        bytes32 _ism,
+        bytes memory _body,
+        bytes memory _hookMetadata
+    ) private returns (bytes32) {
+        require(_router != bytes32(0), "no router specified for destination");
+        emit RemoteCallDispatched(_destination, msg.sender, _router, _ism);
```
Comment (yorhodes) [inline solidity/contracts/middleware/InterchainAccountRouter.sol:561]: I would prefer to have only one `mailbox.dispatch` and `emit RemoteCallDispatched` in the contract
can we refactor slightly?

File: solidity/test/InterchainAccountRouter.t.sol [hunk 5]
```diff
@@ -92,6 +92,9 @@ contract InterchainAccountRouterTest is Test {
     function setUp() public {
         environment = new MockHyperlaneEnvironment(origin, destination);
 
+        igp = new TestInterchainGasPaymaster();
+        gasPaymentQuote = igp.getDefaultGasUsage() * igp.gasPrice();
```
Comment (yorhodes) [inline solidity/test/InterchainAccountRouter.t.sol:96]: why not just use the quote fn?

File: solidity/test/InterchainAccountRouter.t.sol [hunk 6]
```diff
@@ -124,80 +129,102 @@ contract InterchainAccountRouterTest is Test {
...
-        assertEq(ica.owner(), address(destinationRouter));
+    function testFuzz_constructor(address _localOwner) public {
+        OwnableMulticall _account = destinationRouter
+            .getDeployedInterchainAccount(
+                origin,
+                _localOwner,
+                address(originRouter),
+                address(environment.isms(destination))
+            );
+        assertEq(_account.owner(), address(destinationRouter));
```
Comment (yorhodes) [inline solidity/test/InterchainAccountRouter.t.sol:140]: im surprised this works?

File: solidity/test/InterchainAccountRouter.t.sol [hunk 7]
```diff
@@ -124,80 +129,102 @@ contract InterchainAccountRouterTest is Test {
...
+        address _account = originRouter.getRemoteInterchainAccount(
+            address(_localOwner),
+            address(destinationRouter),
+            _ism
         );
         originRouter.enrollRemoteRouterAndIsm(
             destination,
             routerOverride,
-            ismOverride
+            TypeCasts.addressToBytes32(_ism)
```
Comment (yorhodes) [inline solidity/test/InterchainAccountRouter.t.sol:155]: why is the ISM bytes32?

File: solidity/test/InterchainAccountRouter.t.sol [hunk 8]
```diff
@@ -268,89 +307,195 @@ contract InterchainAccountRouterTest is Test {
...
+            ismOverride
+        );
+        uint256 balanceBefore = address(this).balance;
+
+        // act
+        originRouter.callRemote{value: gasLimit * igp.gasPrice() + overpayment}(
+            destination,
+            getCalls(data),
+            metadata
+        );
```
Comment (yorhodes) [inline solidity/test/InterchainAccountRouter.t.sol:420]: should we test underpayment with hook metadata too? as it stands these arent really unit tests

**Review by yorhodes — APPROVED** [review]
thanks for adding fuzzing
some small asks around preserving the simplest interface

File: solidity/contracts/middleware/InterchainAccountRouter.sol (hunk 4 above)
Comment (aroralanuk) [reply solidity/contracts/middleware/InterchainAccountRouter.sol:561]: I cannot put empty metadata so we need an if else regardless of having just one function call mailbox.dispatch

File: solidity/test/InterchainAccountRouter.t.sol (hunk 7 above)
Comment (aroralanuk) [reply solidity/test/InterchainAccountRouter.t.sol:155]: no pointed out yourself, this is remote ism

File: solidity/test/InterchainAccountRouter.t.sol (hunk 6 above)
Comment (aroralanuk) [reply solidity/test/InterchainAccountRouter.t.sol:140]: why?

File: solidity/contracts/middleware/InterchainAccountRouter.sol (hunk 2 above)
Comment (aroralanuk) [reply solidity/contracts/middleware/InterchainAccountRouter.sol]: done

File: solidity/contracts/middleware/InterchainAccountRouter.sol (hunk 3 above)
Comment (aroralanuk) [reply solidity/contracts/middleware/InterchainAccountRouter.sol]: done

File: solidity/test/InterchainAccountRouter.t.sol (hunk 5 above)
Comment (aroralanuk) [reply solidity/test/InterchainAccountRouter.t.sol]: done

**Pull request merged**